core/src/main/java/io/substrait/relation/ProtoRelConverter.java

```diff
@@ -0,0 +1,251 @@
+package io.substrait.relation;
+
+import io.substrait.expression.AggregateFunctionInvocation;
+import io.substrait.expression.Expression;
+import io.substrait.expression.ImmutableExpression;
+import io.substrait.expression.proto.FunctionLookup;
+import io.substrait.expression.proto.ProtoExpressionConverter;
+import io.substrait.function.SimpleExtension;
+import io.substrait.proto.AggregateFunction;
+import io.substrait.proto.AggregateRel;
+import io.substrait.proto.FetchRel;
+import io.substrait.proto.FilterRel;
+import io.substrait.proto.JoinRel;
+import io.substrait.proto.ProjectRel;
+import io.substrait.proto.ReadRel;
+import io.substrait.proto.SortField;
+import io.substrait.proto.SortRel;
+import io.substrait.type.ImmutableNamedStruct;
+import io.substrait.type.NamedStruct;
+import io.substrait.type.Type;
+import io.substrait.type.proto.FromProto;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Converts from proto to pojo rel representation
+ * TODO:
+ * AdvancedExtension
+ * Remap
+ * Missing Rel subclasses: CrossJoin, Set, etc
+ *
+ */
+public class ProtoRelConverter {
+    static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(ProtoRelConverter.class);
+
+    private final FunctionLookup lookup;
+    private final SimpleExtension.ExtensionCollection extensions;
+    private final Type rootType;
+
+    public ProtoRelConverter(FunctionLookup lookup, SimpleExtension.ExtensionCollection extensions, Type rootType) {
+        this.lookup = lookup;
+        this.extensions = extensions;
+        this.rootType = rootType;
+    }
+
+    public Rel from(io.substrait.proto.Rel rel) {
+        io.substrait.proto.Rel.RelTypeCase relType = rel.getRelTypeCase();
+        switch (relType) {
+            case READ -> {
+                ReadRel readRel = rel.getRead();
+                if (readRel.hasVirtualTable()) {
```

Comment:
If you're avoiding nesting for all the other rels, it feels weird to do the nesting for readrel inline.